roll/__init__.py

```diff
@@ -325,9 +337,8 @@ def data_received(self, data: bytes):
             self.response.status = HTTPStatus.BAD_REQUEST
             self.response.body = b'Unparsable request'
             self.write()
-
-    def connection_made(self, transport):
-        self.writer = transport
+        except HttpParserUpgrade:
+            self.upgrade = self.request.headers.get('Upgrade')
```

Comment:
No, it's just a way to have the requested upgrade somewhere. It's not used yet. the "Upgrade" header should contain "websocket" or "keep-alive, websocket", in case of a websocket. We should test it somewhere, probably in the proxy itself, but this wasn't discussed with you or Yohan, so maybe we can take a decision here.

Reply:
I'm not sure I get it, how is the request upgraded then?

Can you add a test with `"keep-alive, websocket"` at that point or is it useless until we actually use it?

Reply:
I got confused in my comment, i meant the upgrade can contain a list of protocols, just like sub-protocols can be a list of subprotocol, separated with commas, ordered by preference.
I added tests and comments in the code and a RFC link, to be less messy than I am in my explanations. Sorry about that :-1: 

Right now, I changed to write the info on the request. That makes much more sense than on the protocol: that was really silly.

It is not really parsed, just read as it is. We trust the handler/proxy to do what is right with it, as I did on the websocket one (fairly naively).